src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -565,6 +566,10 @@ private function createContextListener(ContainerBuilder $container, string $cont
         $listenerId = 'security.context_listener.'.\count($this->contextListeners);
         $listener = $container->setDefinition($listenerId, new ChildDefinition('security.context_listener'));
         $listener->replaceArgument(2, $contextKey);
+        if (null !== $firewallEventDispatcherId) {
+            $listener->replaceArgument(4, new Reference($firewallEventDispatcherId));
+            $listener->addTag('kernel.event_listener', ['event' => KernelEvents::RESPONSE, 'method' => 'onKernelResponse']);
+        }
```

Comment:
Can you explain this piece? Currently, when `ContextListener::authenticate()` is called, that class subscribes to the `KernelEvents:RESPONSE` event. I'm actually not sure what the reason is for that odd setup.

I believe you added this code because we are NOW passing in the firewall-specific dispatcher into `ContextListener`, and not the *main* dispatcher. And so, subscribing to `KernelEvents:RESPONSE` no longer has any effect. That makes sense. Two questions though:

1) Why do we have the current, weird setup where we register the listener inside the class? If that is there for a legitimate reason, then why is it ok for us to NOW register it in a more traditional way?

2) Should (can we?) deprecate the old functionality of having `ContextListener` register the listener? It's a bit odd (and just ugly code to keep around) for `ContextListener` to register the listener on the firewall dispatcher... which will never actually have any effect.

Reply:
See the discussion in a previous comment on this PR: https://github.com/symfony/symfony/pull/40145#discussion_r574802604

Basically, we must only do something in `onKernelResponse()` if `authenticate()` was called for the same request (see referenced PRs in the previous link). Previously, that was "solved" by dynamically registering this listener. For performance reasons, and to make this less weird, I've refactored it to instead rely on saving the state in the request attributes.

Reply:
> Should (can we?) deprecate the old functionality of having ContextListener register the listener?

I don't think we can check whether `onKernelResponse()` listener was registered by the `authenticate()` method or config. So I decided to just always register it in the 5.x, and then remove that line in 6.0. That's a very small hard break (if someone used the `ContextListener` together with the EventDispatcher component outside the framework), but as it is a break, doing it in 6.0 seems better to me.